tools/build_tar/buildtar.go

```diff
@@ -330,7 +328,107 @@ func (f *tarFile) addTar(toAdd string) error {
 }
 
 func (f *tarFile) addDeb(toAdd string) error {
-	return fmt.Errorf("addDeb unimplemented")
+	pkgDataFound, pkgMetadataFound := false, false
+	file, err := os.Open(toAdd)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+	tr := ar.NewReader(bufio.NewReader(file))
+	for {
+		header, err := tr.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		basename := strings.SplitN(header.Name, ".", 2)[0]
+		// BSD variant of ar has file member name terminated with '/'
+		// Source https://www.unix.com/man-page/opensolaris/3HEAD/ar.h/
+		name := strings.TrimRight(header.Name, "/")
+		switch basename {
+		case "data":
+			pkgDataFound = true
+			err = f.addTar(tr, name)
+			if err != nil {
+				return err
+			}
+		case "control":
+			pkgMetadataFound = true
+			f.addDebPackageMetadata(tr, name)
+		}
+	}
+	if !pkgDataFound {
+		return fmt.Errorf("%s does not contain a data file", toAdd)
+	}
+	if !pkgMetadataFound {
+		return fmt.Errorf("%s does not contain a control file", toAdd)
+	}
+	return nil
+}
+
+func (f *tarFile) addDebPackageMetadata(r io.Reader, toAdd string) error {
+	r, err := decompress(r, toAdd)
+	if err != nil {
+		return err
+	}
+	tr := tar.NewReader(r)
+	controlArchiveFound := false
+	var buf bytes.Buffer
+	var header *tar.Header
+	for {
+		header, err = tr.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		if filepath.Base(header.Name) == "control" {
+			controlArchiveFound = true
+			io.Copy(&buf, tr)
+			break
+		}
+
+	}
+	if !controlArchiveFound {
+		return fmt.Errorf("%s does not contain control file", toAdd)
+	}
+	metadata := buf.String()
+	destinationFile := filepath.Join(dpkgStatusDir, extractDebPackageName(metadata, toAdd))
+	header.Name = destinationFile
+	if err := f.tw.WriteHeader(header); err != nil {
```

Comment:
you probably need to call `makeDirs` first, in case the `pkgStatusDir` doesn't exist yet, similar to https://github.com/kubernetes/repo-infra/pull/88.

Reply:
done